File: Task1B/premadelunch.cpp
#include "premadelunch.h"

// Default constructor
PremadeLunch::PremadeLunch() {
    this->perishable = false;
    this->ingredients = "";
    this->allergens = "";
}

// Parameterised constructor
PremadeLunch::PremadeLunch(string name, int price, Supplier* supplier, int volume, int productID, double weight, string expirationDate, bool perishable, string ingredients, string allergens) : Food(name, price, supplier, volume, productID, weight, expirationDate) {
    this->perishable = perishable;
    this->ingredients = ingredients;
    this->allergens = allergens;
}

// Getters and setters
bool PremadeLunch::getPerishable() {
    return this->perishable;
}

string PremadeLunch::getIngredients() {
    return this->ingredients;
}

string PremadeLunch::getAllergens() {
    return this->allergens;
}

void PremadeLunch::setPerishable(bool perishable) {
    this->perishable = perishable;
}

void PremadeLunch::setIngredients(string ingredients) {
    this->ingredients = ingredients;
}

void PremadeLunch::setAllergens(string allergens) {
    this->allergens = allergens;
}

// Display product information
void PremadeLunch::display_product_info() {
    Food::display_product_info();
    cout << "Perishable: " << this->perishable << endl;
    cout << "Ingredients: " << this->ingredients << endl;
    cout << "Allergens: " << this->allergens << endl;
}

// Calculate discount
double PremadeLunch::calculate_discount(string month, int quantity) {
    if (month == "January" || month == "February" || month == "March" || quantity > 10) {
        return getPrice() * 0.5;
    }
    else {
        return getPrice() * 0.1;
    }
}

// Save product information
void PremadeLunch::save(ostream& out) const {
    out << "premadeLunch" << "\n";
    Food::save(out);
    out << perishable << "," << ingredients << "," << allergens << "\n";
}

// Load product information
void PremadeLunch::load(istream& in) {
    Food::load(in);

    string line;
    getline(in, line, '\n');
    stringstream ss(line);

    string token;
    getline(ss, token, ',');
    perishable = stoi(token);

    getline(ss, token, ',');
    ingredients = token;

    getline(ss, token, ',');
    allergens = token;
}

File: Task2A_B/task2a_b/src/main/java/com/cps2004/PremadeLunch.java
package com.cps2004;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ObjectNode;

public class PremadeLunch extends Food {
    private boolean perishable;
    private String ingredients;
    private String allergens;

    // Default constructor
    public PremadeLunch() {
        super();
        this.perishable = false;
        this.ingredients = "";
        this.allergens = "";
    }

    // Parameterized constructor
    public PremadeLunch(boolean perishable, String ingredients, String allergens, String name, int id, Supplier supplier, int price, int volume, double weight, String expirationDate) {
        super(name, id, supplier, price, volume, weight, expirationDate);
        this.perishable = perishable;
        this.ingredients = ingredients;
        this.allergens = allergens;
    }

    // Getters and setters
    public boolean getPerishable() { return perishable;}
    public String getIngredients() { return ingredients;}
    public String getAllergens() { return allergens;}
    public void setPerishable(boolean perishable) { this.perishable = perishable;}
    public void setIngredients(String ingredients) { this.ingredients = ingredients;}
    public void setAllergens(String allergens) { this.allergens = allergens;}

    // Display premade lunch information
    public void display_product_info() {
        super.display_product_info();
        System.out.println("Product perishable: " + perishable);
        System.out.println("Product ingredients: " + ingredients);
        System.out.println("Product allergens: " + allergens);
    }

    @Override
    // Calculate discounted price
    public double calculate_discount(String month, int quantity) {
        if (month.equals("January") || month.equals("February") || month.equals("March") || quantity > 10) {
            return getPrice() * 0.5;
        }
        else {
            return getPrice() * 0.1;
        }
    }

    // Save product information to file
    public void save(ObjectNode newItem) {
        newItem.put("type", "premadeLunch");
        super.save(newItem);
        newItem.put("perishable", perishable);
        newItem.put("ingredients", ingredients);
        newItem.put("allergens", allergens);
    }

    // Load product information from file
    public void load(JsonNode node) {
        super.load(node);
        this.perishable = node.get("perishable").asBoolean();
        this.ingredients = node.get("ingredients").asText();
        this.allergens = node.get("allergens").asText();
    }
}
